hare/eligibility/oracle.go

```diff
@@ -249,58 +269,68 @@ func (o *Oracle) Proof(ctx context.Context, layer types.LayerID, round int32) ([
 	// lock until any return
 	// note: no need to lock per safeEp - we do not expect many concurrent requests per safeEp (max two)
 	o.lock.Lock()
 	defer o.lock.Unlock()
 
-	// check cache
-	if val, exist := o.activesCache.Get(safeEp); exist {
+	// we can't read blocks during genesis epochs as there are none
+	if targetLayer.GetEpoch().IsGenesis() {
+		return nil, errGenesis
+	}
+
+	// check cache first
+	if val, exist := o.activesCache.Get(targetLayer.GetEpoch()); exist {
 		return val.(map[string]struct{}), nil
 	}
 
-	// build a map of all blocks on the current layer
-	mp, err := o.blocksProvider.ContextuallyValidBlock(sl)
+	// we first try to get the hare active set for a range of safe layers: start with the set of active blocks
+	safeLayerStart, safeLayerEnd := safeLayerRange(
+		targetLayer, types.LayerID(o.cfg.ConfidenceParam), types.LayerID(o.layersPerEpoch), types.LayerID(o.cfg.EpochOffset))
+	activeBlockIDs := make(map[types.BlockID]struct{})
+	for layerID := safeLayerStart; layerID <= safeLayerEnd; layerID++ {
+		layerBlockIDs, err := o.meshdb.LayerContextuallyValidBlocks(layerID)
+		if err != nil {
+			return nil, fmt.Errorf("error getting active blocks for layer %v for target layer %v: %w",
+				layerID, targetLayer, err)
+		}
+		for blockID := range layerBlockIDs {
+			activeBlockIDs[blockID] = struct{}{}
+		}
+	}
+
+	// now read the set of ATXs referenced by these blocks
+	hareAtxs, err := o.atxdb.ActiveSetFromBlocks(safeLayerStart.GetEpoch(), activeBlockIDs)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error getting ATXs for target layer %v: %w", targetLayer, err)
 	}
 
-	// no contextually valid blocks
-	if len(mp) == 0 {
-		o.With().Error("could not calculate hare active set size: no contextually valid blocks",
-			layer,
-			layer.GetEpoch(),
-			log.FieldNamed("safe_layer_id", sl),
-			log.FieldNamed("safe_epoch_id", safeEp))
-		return nil, errNoContextualBlocks
+	if len(hareAtxs) > 0 {
+		o.activesCache.Add(targetLayer.GetEpoch(), hareAtxs)
+		return hareAtxs, nil
 	}
 
-	activeMap, err := o.getActiveSet(safeEp-1, mp)
-	if err != nil {
-		o.With().Error("could not retrieve active set size",
-			log.Err(err),
-			layer,
-			layer.GetEpoch(),
-			log.FieldNamed("safe_layer_id", sl),
-			log.FieldNamed("safe_epoch_id", safeEp))
-		return nil, err
+	// if we failed to get a Hare active set, we fall back on reading the Tortoise active set targeting this epoch
+	atxs := o.atxdb.GetEpochAtxs(targetLayer.GetEpoch() - 1)
+	if len(atxs) == 0 {
+		return nil, fmt.Errorf("empty active set for layer %v in non-genesis epoch %v",
+			targetLayer, targetLayer.GetEpoch())
 	}
 
-	// update
-	o.activesCache.Add(safeEp, activeMap)
+	// extract the nodeIDs
+	activeMap := make(map[string]struct{}, len(atxs))
+	for _, atxid := range atxs {
+		atx, err := o.atxdb.GetAtxHeader(atxid)
+		if err != nil {
+			return nil, fmt.Errorf("error getting ATX %v for target layer %v: %w", atxid, targetLayer, err)
```

Comment:
Because we got this `atxid` from the database (vs. some incoming object) I think it's important to note that this is a consistency issue in the database. This should never happen unless we have a bug somewhere, and we may want to add some kind of monitoring for this specific keyword in errors.

```suggestion
			return nil, fmt.Errorf("inconsistent state: error getting ATX %v for target layer %v: %w", atxid, targetLayer, err)
```

Reply:
made this change